### PR TITLE
feat(NcAvatar): implement custom javascript hook action for the contacts menu

### PR DESCRIPTION
**Pull request by st3iny**
### ☑️ Resolves

- Part of: https://github.com/nextcloud/server/issues/44776
- Server PR: https://github.com/nextcloud/server/pull/49375

### 📒 Description 

Adds a new contacts menu that is able to trigger hooks registered via `registerContactsMenuAction("foo-bar")`.

```typescript
export interface ContactsMenuAction {
	id: string,
	displayName: (entry: ContactsMenuEntry) => string,
	enabled: (entry: ContactsMenuEntry) => boolean,
	iconSvg: (entry: ContactsMenuEntry) => string,
	callback: (entry: ContactsMenuEntry) => void,
}
```

### 🖼️ Screenshots

Nothing really to show here. Please have a look [here](https://github.com/nextcloud/calendar/pull/6502) for an implementation.

### 🏁 Checklist

- [ ] ⛑️ Tests are included or are not applicable
- [ ] 📘 Component documentation has been extended, updated or is not applicable
- [x] 3️⃣ Backport to `next` requested with a Vue 3 upgrade

### CODE REVIEW

File: src/functions/contactsMenu/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+type ContactsMenuCallback = () => void
+
+interface ContactsMenuHookProps {
+	id: string;
+	callback: ContactsMenuCallback;
+}
+
+declare global {
+	interface Window {
+		_vue_contacts_menu_hooks: Record<string, ContactsMenuHookProps>;
```
Comment (susnux) [inline src/functions/contactsMenu/index.ts:15]: naming should be consistent with other so better `_nc_...`

File: src/functions/contactsMenu/index.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+type ContactsMenuCallback = () => void
+
+interface ContactsMenuHookProps {
+	id: string;
+	callback: ContactsMenuCallback;
+}
+
+declare global {
```
Comment (susnux) [inline src/functions/contactsMenu/index.ts:13]: Might end up in public types, so this global declarations should be in a `src/globals.d.ts` file and not something we export

File: src/functions/contactsMenu/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	interface Window {
+		_vue_contacts_menu_hooks: Record<string, ContactsMenuHookProps>;
+	}
+}
+
+if (!window._vue_contacts_menu_hooks) {
+	window._vue_contacts_menu_hooks = {}
+}
+
+const registerContactsMenuHook = (id: string, callback: ContactsMenuCallback) => {
```
Comment (susnux) [inline src/functions/contactsMenu/index.ts:23]: I think for our syntax it is more common to directly export.
Also using `function` will hoist the declaration of it which prevents issues in general.

```suggestion
export function registerContactsMenuHook(id: string, callback: ContactsMenuCallback): void {
```

Comment (susnux) [inline src/functions/contactsMenu/index.ts:23]: See above about the side effects, just move it to this function will prevent side effects of importing this module so tree shaking can remove all imports if not used and do keep not imports for side effects only.

```suggestion
const registerContactsMenuHook = (id: string, callback: ContactsMenuCallback) => {
    window._vue_contacts_menu_hooks = window._vue_contacts_menu_hooks ?? {}
```

File: src/functions/contactsMenu/index.ts (hunk 1 above)
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:34]: Done

File: src/functions/contactsMenu/index.ts (hunk 2 above)
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:32]: Ok, I'll move the global `Window._nc_[...]` definition and interface ` ContactsMenuHookProps` there.

File: src/functions/contactsMenu/index.ts (hunk 3 above)
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:42]: Yeah, I also like exporting directly and not having const closures more.

Done.

File: src/functions/contactsMenu/index.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	callback: ContactsMenuCallback;
+}
+
+declare global {
+	interface Window {
+		_vue_contacts_menu_hooks: Record<string, ContactsMenuHookProps>;
+	}
+}
+
+if (!window._vue_contacts_menu_hooks) {
```
[the remark this answers is not in the document]
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:38]: Done

File: src/functions/contactsMenu/index.ts (hunk 3 above)
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:42]: Done

File: src/functions/contactsMenu/index.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	window._vue_contacts_menu_hooks = {}
+}
+
+const registerContactsMenuHook = (id: string, callback: ContactsMenuCallback) => {
+	if (window._vue_contacts_menu_hooks[id]) {
+		console.error(`Hook for id ${id} already registered`)
```
[the remark this answers is not in the document]
Comment (st3iny) [reply src/functions/contactsMenu/index.ts:44]: TIL, that we also have a logger in nc-vue. I refactored the code accordingly.

File: src/functions/contactsMenu/index.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	window._vue_contacts_menu_hooks[id] = {
+		id,
+		callback,
+	}
+}
+
+const callContactsMenuHook = (id: string) => {
+	window._vue_contacts_menu_hooks[id]?.callback()
```
[the remark this answers is not in the document]
Comment (st3iny) [reply src/functions/contactsMenu/index.ts]: Done

File: src/functions/contactsMenu/index.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Register a contacts and avatar menu action that will invoke the given callback on click.
+ *
+ * @param {ContactsMenuAction} action The action to register
+ */
+export function registerContactsMenuAction(action: ContactsMenuAction): void {
+	window._nc_contacts_menu_hooks ??= {}
+
+	if (window._nc_contacts_menu_hooks[action.id]) {
+		logger.error(`ContactsMenu action for id ${action.id} was already registered`, {
```
Comment (ChristophWurst) [inline src/functions/contactsMenu/index.ts:44]:
```suggestion
		logger.error(`ContactsMenu action for id ${action.id} has already been registered`, {
```

**Review by ChristophWurst — APPROVED** [review]
Looks good

File: src/functions/contactsMenu/index.ts (hunk 7 above)
Comment (st3iny) [reply src/functions/contactsMenu/index.ts]: Done

File: src/functions/contactsMenu/index.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+	if (window._nc_contacts_menu_hooks[action.id]) {
+		logger.error(`ContactsMenu action for id ${action.id} has already been registered`, {
+			action,
+		})
+		return
+	}
+
+	window._nc_contacts_menu_hooks[action.id] = action
+}
```
Comment (skjnldsv) [inline src/functions/contactsMenu/index.ts:51]: that makes it not available for other repos ?
Isn't there a library that would be fit to receive that API ?

Comment (skjnldsv) [reply src/functions/contactsMenu/index.ts:51]: Hilarious, I thought I was in /server :eyes:

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**